torch/_refs/__init__.py

```diff
@@ -4010,41 +4010,18 @@ def linspace(
         "requires_grad": requires_grad,
     }
     if steps == 0:
-        ret = torch.full((0,), 0, dtype=dtype, **factory_kwargs)  # type: ignore[call-overload]
+        return torch.full((0,), 0, dtype=dtype, **factory_kwargs)  # type: ignore[call-overload]
     elif steps == 1:
-        ret = torch.full((1,), start, dtype=dtype, **factory_kwargs)  # type: ignore[call-overload]
+        return torch.full((1,), start, dtype=dtype, **factory_kwargs)  # type: ignore[call-overload]
     elif start == end:
-        ret = torch.full((steps,), start, dtype=dtype, **factory_kwargs)  # type: ignore[call-overload]
+        return torch.full((steps,), start, dtype=dtype, **factory_kwargs)  # type: ignore[call-overload]
     else:
-        if prims.utils.is_integer_dtype(dtype):
-            # We need to cast to int, so to avoid off-by-one issues
-            # do the entire computation with ints when we can
-            assert isinstance(start, int) and isinstance(end, int)
-            step_size_x_denom = end - start
-            eps = 1 if end > start else -1
-            denom = steps - 1
-            ret = prims.to_dtype(
-                torch.arange(
-                    start * denom,
-                    end * denom + eps,
-                    step_size_x_denom,
-                    dtype=torch.int64,
-                    **factory_kwargs,  # type: ignore[arg-type]
-                )
-                / denom,
-                dtype,
-            )
-        else:
-            step_size = (end - start) / (steps - 1)
-            eps = step_size / 2
-            ret = prims.to_dtype(
-                torch.arange(  # type: ignore[call-overload]
-                    start, end + eps, step_size, dtype=torch.float64, **factory_kwargs
-                ),
-                dtype,
-            )
-
-    return ret
+        step_size = (end - start) / (steps - 1)
+        eps = step_size / 2
+        ret_double = torch.arange(  # type: ignore[call-overload]
```

Comment:
And it doesn't directly answer the question of what to do when start and step are integer (yes yes I know that integer is real, nevertheless, it's usual and customary to explicitly distinguish between them).

Reply:
alas, this op cannot be computed natively (for all I know) in those ops without a casting. Even more, this op is not well-defined on ints. IMO, ints shouldn't be supported on this op, and these things are what we get for supporting them...

Reply:
Now, given the current state of things, I think we get to decide what's the definition, and it can very well be "compute things on real floating precision and then cast down".

Reply:
The current decomposition on double yields (at least locally) the same results than the one that was there before, so I think this change is fine.

I think this is as good as it gets with this decomposition and those annoying inputs really, unless you have a different proposal.

Reply:
as @mruberry suggested above, and per docs, the computation type should be global default type, not double.